File: app/src/test/java/com/allan/androidlearning/SimpleEncodeUtil.java
package com.allan.androidlearning;

class SimpleEncodeUtil {
    private final String original;
    private final char startChar;
    private final int[] indexes;
 
    public SimpleEncodeUtil(String original, char startChar, int[] indexes) { 
        this.original  = original; 
        this.startChar  = startChar; 
        this.indexes  = indexes; 
    } 
 
    public String decode() { 
        StringBuilder sb = new StringBuilder(); 
        for (int index : indexes) { 
            char ch = (char) (startChar + index); 
            sb.append(ch);  
        } 
        return sb.toString();  
    } 
 
    private String toClassCreatorString() { 
        StringBuilder sb = new StringBuilder("intArrayOf("); 
        for (int index : indexes) { 
            sb.append(index).append(',');  
        } 
        sb.append(')');  
        String str = "val originalStr = SimpleEncodeUtil(\"\", '%s', %s).decode() // originalString: %s"; 
        return String.format(str,  "" + startChar, sb.toString(),  original); 
    } 
 
    public static String originalToStringSplitClassString(String originalStr) { 
        char[] chars = originalStr.toCharArray();  
        char[] rands = {'0', '1', '2', '3', '4', '5'}; 
        char startChar = rands[(int) (Math.random()  * rands.length)];  
        int[] indexes = new int[chars.length]; 
        for (int i = 0; i < chars.length;  i++) { 
            indexes[i] = chars[i] - startChar; 
        } 
        return new SimpleEncodeUtil(originalStr, startChar, indexes).toClassCreatorString(); 
    } 
 
    public static void main(String[] args) { 
        String originalStr = "com.ss.android.lark";  
        String classStr = SimpleEncodeUtil.originalToStringSplitClassString(originalStr);  
        System.out.println(classStr);  
    }
} 

File: test/src/test/java/com/example/test/SimpleEncodeUtil.kt
package com.example.test

class SimpleEncodeUtil {
    class Encoder(private val origStr:String) {
        fun originalToStringSplitClassString() : String {
            val chars = origStr.toCharArray()
            val rands = arrayOf(1, 2, 3, 4, 5)
            val offset = rands[(Math.random() * rands.size).toInt()]
            val indexes = mutableListOf<Int>()
            indexes.add(offset + 100)
            for (ch in chars) {
                val index = ch - offset
                indexes.add(index.code)
            }
            return toClassCreatorString(indexes.toIntArray())
        }

        private fun toClassCreatorString(indexes: IntArray): String {
            val sb = StringBuilder("intArrayOf(")
            for (index in indexes) {
                sb.append(index).append(',')
            }
            sb.append(')')
            val str = " // originalString: %s\nval originalStr = SimpleEncodeUtil.Decoder().decode(%s)"
            val str2 = "// originalString: %s\nAppNative.simpleDecoder(%s)"
            return String.format(str, origStr, sb.toString()) + "\n" + String.format(str2, origStr, sb.toString())
        }
    }

    class Decoder {
        //不校验indexes的数组是否为空
        fun decode(indexes: IntArray) : String {
            val sb = StringBuilder()
            val offset = indexes[0] - 100
            var i = 1
            while(i < indexes.size) {
                val ch = (indexes[i] + offset).toChar()
                sb.append(ch)
                i++
            }
            return sb.toString()
        }
     }
}